content/posts/Docker Tags.md

```diff
@@ -0,0 +1,54 @@
+
+So maybe you were using `FROM python:3` and one day your builds started failing when the version jumped from `3.7` to `3.8`. You then decide to switch to using `FROM python:3.7` instead. Now, when 3.8, 3.9, 3.10 etc. are released, you won’t automatically bump to those minor release versions without you changing it in your Dockerfile.
+
+You can begin to see where this pattern goes. By starting out at the least specific version tag, you are basically automatically going to be building the newest and latest version of that image. Moving down the spectrum, you can specify a major release versions to use, or the major + minor release version, or even the major + minor + security patch level to use. You also have one additional option: specifying the exact `sha256` fingerprint of an image.
+
+Instead of numbered version tag, you could specify `FROM python@sha256:7e6c00cc553fdce06c1bcfcbf34c73a0f3623a8fc9ce88c8fb91671f2cbc0dab` in your Dockerfile. This is the exact fingerprint of a specific image, and it will never change. This would ensure that nothing about your base image would ever change out from underneath you. This means you probably won't see as many unexpected build failures when some downstream packages. The tradeoff is you won’t be pulling in any newer versions of base images or packages that contain updates, like security patches.
+
+To sum is up: When using something like the `:latest` tag of a public Docker image you are basically always going to be running a patched version of that container, but the packages inside will constantly be changing. By using a specific `sha256` the image won’t change, meaning you won’t get any patches or updates without updating the tag in your Dockerfile.
+
+So which strategy should you use? Ultimately this should be a discussion you will want to have with anyone working on the codebase. Depending on a number of factors such as the number of people working on an application, the resources the team has available, the complexity of the application, and the language used, one tagging strategy may be more appropriate over another.
+
+If you are looking for a place to start out, try starting with using the minor stable release of the language you are using, such as `python:3.7`. Or in the case of ruby, try something like `ruby:2.6`. For Go, start with something like `golang:1.12`, (or try the `alpine` or `slim` versions if you want to use much smaller images). If you run into some unexpected behavior in testing or are constantly having build failures from dependencies changing out from underneath you, move to a more specific tag. You may end up deciding that for that particular Dockerfile you want to use the exact `sha256` for your base container. This is a fine option, just remember that it means you’ll also need to regularly review and update your base image manually in order to pull in any security patches.
```

Comment:
I think this is a good compromise. Is it worth teasing this out into two different use cases: third-party software from upstream that you don't control, and your own builds of your own software. You might want different strategies for each.